### PR TITLE
ISSUE 3850 - Move notifications to internal DB

### LINKED ISSUE #3850: Move notifications into a single collection inside internal db
### Description
At the moment notifications are stored inside a `notifications` db, with each user having their own collection.

As part of the effort to reduce the number of namespaces we use inside mongo, we want to merge them all into a single collection inside `internal`, similar to what we've done with login records (#3694)

Some changes to be made to the records:
- we will need to add the `user` into the record as this is currently deduced by the collection name
- Whilst we're at it, we should change the timestamp to Date type
image

This work will need a migration script to move all existing notifications into the new collection, ensure all new notifications are stored in the same way, and ensure all `GET` requests are fetching records from the right place.

We will also need to look into indexing the query (most likely by username) as we will have a lot of data in a single collection.

### Goals
- write a migration script to migrate the data into `internal.notifications`
- ensure index is applied to this collection
- update code to write/read from the new place

### PR DESCRIPTION
**Pull request by Charence**
This fixes #3850 

#### Description
- Migration script to move existing notifications
- Create indices
- Backend updated to read/write notifications from notifications DB

#### Test cases
- Should work as before

### CODE REVIEW

**Review by carmenfan — CHANGES_REQUESTED** [review]
I don't know where it needs to be, but you need something to convert the `Date` back to timestamp when we send it out as a response to an endpoint

File: backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+const { logger } = require(`${v5Path}/utils/logger`);
+const { initialise } = require(`${v5Path}/models/notifications`);
+
+const INTERNAL_DB = 'internal';
+const NOTIFICATIONS_DB = 'notifications';
+const NOTIFICATIONS_COLL = 'notifications';
+
+const processCollection = async (user) => {
+	const migrationOps = (await find(NOTIFICATIONS_DB, user, {})).map((entry) => ({
+		updateOne: {
```
[the remark this answers is not in the document]
Comment (Charence) [reply backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js:30]: I did this as I ran the script multiple times during testing. I can change it `insertMany` if we're only going to run it once.

File: backend/src/v4/models/notification.js [hunk 2]
```diff
@@ -246,15 +248,16 @@ module.exports = {
...
+		await db.updateMany(INTERNAL_DB, NOTIFICATIONS_COLL, { user }, { $set: data });
 	},
 
 	/**
 	 * This delete all notifications for the particular user
 	 */
-	deleteAllNotifications: async function(username) {
-		await db.deleteMany(NOTIFICATIONS_DB, username, {});
+	deleteAllNotifications: async function(user) {
+		// TODO this exists in v5
```
[the remark this answers is not in the document]
Comment (Charence) [reply backend/src/v4/models/notification.js:259]: After discussion last time, I'll remove this comment as we'll handle notifications backend v5 migration separately.

File: backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js:30]: if you do a collection at a time, then you just have to do an insert many and drop collection

so in theory, you shouldn't run into index problems, furthermore, if you do, we shouldn't be doing an update, we should just ignore it (as it means it has been migrated)

File: backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js [hunk 3]
```diff
@@ -26,19 +26,27 @@ const NOTIFICATIONS_DB = 'notifications';
...
-		updateOne: {
-			filter: { _id: entry._id },
-			update: { $set: {
-				...entry,
-				user,
-				timestamp: new Date(entry.timestamp),
-			} },
-			upsert: true,
-		},
+	const updatedNotifications = (await find(NOTIFICATIONS_DB, user, {})).map((entry) => ({
```
Comment (carmenfan) [inline backend/src/scripts/migrations/5.2/moveNotificationsToInternal.js:29]: Probably better to do this
```suggestion
	const updatedNotifications = (await find(NOTIFICATIONS_DB, user, {})).map(({timestamp, ...entry}) => ({
```

File: backend/src/v4/models/notification.js [hunk 4]
```diff
@@ -142,7 +142,7 @@ const fillModelData = async function(fullNotifications) {
 	notifications.forEach (notification => {
 		const teamSpace = modelsData[notification.teamSpace] || {};
 		const {name, federate} = teamSpace[notification.modelId] || {};
-		Object.assign(notification, {modelName: name, federation: federate, project: (modelToProject[notification.teamSpace] ?? {})[notification.modelId]});
+		Object.assign(notification, {timestamp: new Date(notification.timestamp).getTime(), modelName: name, federation: federate, project: (modelToProject[notification.teamSpace] ?? {})[notification.modelId]});
```
Comment (carmenfan) [inline backend/src/v4/models/notification.js:145]: I think you can just do this? Unless i'm missing something there shouldn't be a situation where the timestamp is not a Date object?

```suggestion
		Object.assign(notification, {timestamp: notification.timestamp.getTime(), modelName: name, federation: federate, project: (modelToProject[notification.teamSpace] ?? {})[notification.modelId]});
```

**Review by carmenfan — CHANGES_REQUESTED** [review]

**Review by carmenfan — APPROVED** [review]

**Pull request merged**